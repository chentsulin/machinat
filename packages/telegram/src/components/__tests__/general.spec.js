import Machinat from '@machinat/core';
import Renderer from '@machinat/core/renderer';
import generalComponentDelegator from '../general';

const renderer = new Renderer('messenger', generalComponentDelegator);

test('render shallow elements match snapshot', async () => {
  const results = await Promise.all(
    [
      <p>abc</p>,
      <b>important</b>,
      <i>italic</i>,
      <s>nooooo</s>,
      <u>underlined</u>,
      <code>foo.bar()</code>,
      <pre>foo.bar('hello world')</pre>,
      <br />,
    ].map((ele) => renderer.render(ele))
  );

  expect(results).toMatchSnapshot();
  expect(results.flat().map((seg) => seg.value)).toMatchInlineSnapshot(`
    Array [
      Object {
        "method": "sendMessage",
        "parameters": Object {
          "parse_mode": "HTML",
          "text": "abc",
        },
      },
      "<b>important</b>",
      "<i>italic</i>",
      "<s>nooooo</s>",
      "<u>underlined</u>",
      "<code>foo.bar()</code>",
      "<pre>foo.bar('hello world')</pre>",
      "
    ",
    ]
  `);
});

test('render nested elements match snapshot', async () => {
  const segments = await renderer.render(
    <p>
      Mic test{' '}
      <code>
        Hello, <b>Luke Skywalker!</b>
      </code>
      <br />
      You know what?
      <br />
      <i>
        <u>I'm your</u> <s>FATHER</s> <code>droid</code>.
      </i>
      <br />
      <br />
      <pre>May the force be with you!</pre> Test over
    </p>
  );
  expect(segments).toMatchSnapshot();
  expect(segments.map((seg) => seg.value)).toMatchInlineSnapshot(`
    Array [
      Object {
        "method": "sendMessage",
        "parameters": Object {
          "parse_mode": "HTML",
          "text": "Mic test <code>Hello, <b>Luke Skywalker!</b></code>
    You know what?
    <i><u>I'm your</u> <s>FATHER</s> <code>droid</code>.</i>

    <pre>May the force be with you!</pre> Test over",
        },
      },
    ]
  `);
});

test('<p/> hoist textual children into text message object', async () => {
  const segments = await renderer.render(
    <p>
      foo
      <br />
      bar
      <br />
      <br />
      baz
    </p>
  );

  expect(segments).toMatchSnapshot();
  expect(segments.map((seg) => seg.value)).toMatchInlineSnapshot(`
    Array [
      Object {
        "method": "sendMessage",
        "parameters": Object {
          "parse_mode": "HTML",
          "text": "foo
    bar

    baz",
        },
      },
    ]
  `);
});

test('throw if non-texual value received', async () => {
  const children = (
    <>
      foo
      <img />
      bar
    </>
  );

  await expect(
    renderer.render(<p>{children}</p>)
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"non-textual node <img /> received, only textual nodes and <br/> allowed"`
  );

  for (const node of [
    <b>{children}</b>,
    <i>{children}</i>,
    <s>{children}</s>,
    <u>{children}</u>,
    <code>{children}</code>,
    <pre>{children}</pre>,
  ]) {
    // eslint-disable-next-line no-await-in-loop
    await expect(renderer.render(node)).rejects.toThrow(
      'non-textual node <img /> received, only textual nodes allowed'
    );
  }
});

test('render null if content is empty', async () => {
  const elements = [<b />, <i />, <s />, <u />, <p />, <code />, <pre />];
  for (const element of elements) {
    // eslint-disable-next-line no-await-in-loop
    await expect(renderer.render(element)).resolves.toEqual(null);
  }
});

test('media elements match snapshot', async () => {
  const segments = await renderer.render(
    <>
      <img src="http://avatar.my.bot" />
      <video src="http://vid.my.bot" />
      <audio src="http://sound.my.bot" />
      <file src="http://profile.my.bot" />
    </>
  );
  expect(segments).toMatchSnapshot();
  expect(segments.map((seg) => seg.value)).toMatchInlineSnapshot(`
    Array [
      Object {
        "message": Object {
          "method": "sendPhoto",
          "parameters": Object {
            "photo": "http://avatar.my.bot",
          },
        },
      },
      Object {
        "message": Object {
          "method": "sendVideo",
          "parameters": Object {
            "video": "http://vid.my.bot",
          },
        },
      },
      Object {
        "message": Object {
          "method": "sendAudio",
          "parameters": Object {
            "audio": "http://sound.my.bot",
          },
        },
      },
      Object {
        "message": Object {
          "method": "sendDocument",
          "parameters": Object {
            "document": "http://profile.my.bot",
          },
        },
      },
    ]
  `);
});
