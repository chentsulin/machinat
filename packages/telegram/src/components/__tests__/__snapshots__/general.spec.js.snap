// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<p/> hoist plain text into text message object 1`] = `
Array [
  Object {
    "node": <p>
      foo
      <br />
      bar
      <br />
      <br />
      baz
    </p>,
    "path": "$",
    "type": "unit",
    "value": Object {
      "method": "sendMessage",
      "parameters": Object {
        "parse_mode": "HTML",
        "text": "foo",
      },
    },
  },
  Object {
    "node": <p>
      foo
      <br />
      bar
      <br />
      <br />
      baz
    </p>,
    "path": "$",
    "type": "unit",
    "value": Object {
      "method": "sendMessage",
      "parameters": Object {
        "parse_mode": "HTML",
        "text": "bar",
      },
    },
  },
  Object {
    "node": <p>
      foo
      <br />
      bar
      <br />
      <br />
      baz
    </p>,
    "path": "$",
    "type": "unit",
    "value": Object {
      "method": "sendMessage",
      "parameters": Object {
        "parse_mode": "HTML",
        "text": "baz",
      },
    },
  },
]
`;

exports[`elements match snapshot 1`] = `
Array [
  Object {
    "node": <p>
      abc
    </p>,
    "path": "$::0",
    "type": "unit",
    "value": Object {
      "method": "sendMessage",
      "parameters": Object {
        "parse_mode": "HTML",
        "text": "abc",
      },
    },
  },
  Object {
    "node": <b>
      important
    </b>,
    "path": "$::2",
    "type": "text",
    "value": "<b>important</b>",
  },
  Object {
    "node": <i>
      italic
    </i>,
    "path": "$::4",
    "type": "text",
    "value": "<i>italic</i>",
  },
  Object {
    "node": <s>
      nooooo
    </s>,
    "path": "$::6",
    "type": "text",
    "value": "<s>nooooo</s>",
  },
  Object {
    "node": <u>
      underlined
    </u>,
    "path": "$::8",
    "type": "text",
    "value": "<u>underlined</u>",
  },
  Object {
    "node": <code>
      foo.bar()
    </code>,
    "path": "$::10",
    "type": "text",
    "value": "<code>foo.bar()</code>",
  },
  Object {
    "node": <pre>
      foo.bar('hello world')
    </pre>,
    "path": "$::12",
    "type": "text",
    "value": "<pre>foo.bar('hello world')</pre>",
  },
]
`;

exports[`media elements match snapshot 1`] = `
Array [
  Object {
    "node": <img
      src="http://avatar.my.bot"
    />,
    "path": "$::0",
    "type": "unit",
    "value": Object {
      "message": Object {
        "method": "sendPhoto",
        "parameters": Object {
          "photo": "http://avatar.my.bot",
        },
      },
    },
  },
  Object {
    "node": <video
      src="http://vid.my.bot"
    />,
    "path": "$::1",
    "type": "unit",
    "value": Object {
      "message": Object {
        "method": "sendVideo",
        "parameters": Object {
          "video": "http://vid.my.bot",
        },
      },
    },
  },
  Object {
    "node": <audio
      src="http://sound.my.bot"
    />,
    "path": "$::2",
    "type": "unit",
    "value": Object {
      "message": Object {
        "method": "sendAudio",
        "parameters": Object {
          "audio": "http://sound.my.bot",
        },
      },
    },
  },
  Object {
    "node": <file
      src="http://profile.my.bot"
    />,
    "path": "$::3",
    "type": "unit",
    "value": Object {
      "message": Object {
        "method": "sendDocument",
        "parameters": Object {
          "document": "http://profile.my.bot",
        },
      },
    },
  },
]
`;

exports[`nested elements match snapshot 1`] = `
Array [
  Object {
    "node": <p>
      Mic test
       
      <code>
        Hello, 
        <b>
          Luke Skywalker!
        </b>
      </code>
      <br />
      You know what?
      <br />
      <i>
        <u>
          I'm your
        </u>
         
        <s>
          FATHER
        </s>
         
        <code>
          droid
        </code>
        .
      </i>
      <br />
      <br />
      <pre>
        May the force be with you!
      </pre>
       Test over
    </p>,
    "path": "$",
    "type": "unit",
    "value": Object {
      "method": "sendMessage",
      "parameters": Object {
        "parse_mode": "HTML",
        "text": "Mic test <code>Hello, <b>Luke Skywalker!</b></code>",
      },
    },
  },
  Object {
    "node": <p>
      Mic test
       
      <code>
        Hello, 
        <b>
          Luke Skywalker!
        </b>
      </code>
      <br />
      You know what?
      <br />
      <i>
        <u>
          I'm your
        </u>
         
        <s>
          FATHER
        </s>
         
        <code>
          droid
        </code>
        .
      </i>
      <br />
      <br />
      <pre>
        May the force be with you!
      </pre>
       Test over
    </p>,
    "path": "$",
    "type": "unit",
    "value": Object {
      "method": "sendMessage",
      "parameters": Object {
        "parse_mode": "HTML",
        "text": "You know what?",
      },
    },
  },
  Object {
    "node": <p>
      Mic test
       
      <code>
        Hello, 
        <b>
          Luke Skywalker!
        </b>
      </code>
      <br />
      You know what?
      <br />
      <i>
        <u>
          I'm your
        </u>
         
        <s>
          FATHER
        </s>
         
        <code>
          droid
        </code>
        .
      </i>
      <br />
      <br />
      <pre>
        May the force be with you!
      </pre>
       Test over
    </p>,
    "path": "$",
    "type": "unit",
    "value": Object {
      "method": "sendMessage",
      "parameters": Object {
        "parse_mode": "HTML",
        "text": "<i><u>I'm your</u> <s>FATHER</s> <code>droid</code>.</i>",
      },
    },
  },
  Object {
    "node": <p>
      Mic test
       
      <code>
        Hello, 
        <b>
          Luke Skywalker!
        </b>
      </code>
      <br />
      You know what?
      <br />
      <i>
        <u>
          I'm your
        </u>
         
        <s>
          FATHER
        </s>
         
        <code>
          droid
        </code>
        .
      </i>
      <br />
      <br />
      <pre>
        May the force be with you!
      </pre>
       Test over
    </p>,
    "path": "$",
    "type": "unit",
    "value": Object {
      "method": "sendMessage",
      "parameters": Object {
        "parse_mode": "HTML",
        "text": "<pre>May the force be with you!</pre> Test over",
      },
    },
  },
]
`;
